### PR TITLE
Test scenarios

### LINKED ISSUE #388: Test scenarios
Currently, we check that scenarios can be parsed.

I propose we test that they can be solved:

- add a `solution` field to scenarios, which will contain the winning program
- load this program to `base`
- run CESK machine
- verify the `win` condition

This would help us make sure that scenarios are functional, which will be harder once we have many of them for the tutorial.

It would also create an easy way to write tests without knowing any Haskell!

### PR DESCRIPTION
**Pull request by xsebek**
- add winning solutions to scenarios
- test the winning solutions
  - fail if any robot has a `Fatal error` in its log
  - timeout in X seconds (default 1)
- replace startup logging to `stdout` with marking errors
- fix creative mode ignoring scenario inventory
- fix integration test discovery (challenges are in a subfolder, so recurse)
- add a *Testing* scenario collection hidden without the `--cheat` flag
- closes #388

### CODE REVIEW

**Comment by xsebek** [comment]
@byorgey I wrote a test case for #394, but then I needed `expectFail`, so I put the automatic testing of all scenarios to rest.

I also allow `testSolution'` to check the final `GameState` so automating that is probably not that useful.

**Comment by xsebek** [comment]
I really like [tasty-expected-failure](https://hackage.haskell.org/package/tasty-expected-failure-0.12.3/docs/Test-Tasty-ExpectedFailure.html):
|image|
|-|

**Comment by xsebek** [comment]
Looking at the image above, I noticed we do not actually test all scenarios. :sweat_smile: 

Now it is fixed. :heavy_check_mark:

File: src/Swarm/Game/State.hs [hunk 1]
```diff
@@ -515,7 +522,7 @@ playScenario em scenario userSeed toRun g = do
       & ix 0 . robotInventory
         %~ case scenario ^. scenarioCreative of
           False -> id
-          True -> const (fromElems (map (0,) things))
+          True -> union (fromElems (map (0,) things))
```
Comment (byorgey) [inline src/Swarm/Game/State.hs:525]: What is the purpose of this change?

File: src/Swarm/TUI/Controller.hs [hunk 2]
```diff
@@ -145,7 +145,7 @@ handleNewGameMenuEvent scenarioStack@(curMenu :| rest) s = \case
             & gameState .~ gs'
       Just (SICollection _ c) ->
         continue $
-          s & uiState . uiMenu .~ NewGameMenu (NE.cons (mkScenarioList c) scenarioStack)
+          s & uiState . uiMenu .~ NewGameMenu (NE.cons (mkScenarioList (s ^. gameState . creativeMode) c) scenarioStack)
```
Comment (byorgey) [inline src/Swarm/TUI/Controller.hs:148]: I don't see how the `creativeMode` flag in the `gameState` is relevant here.  It doesn't really tell us much unless we're currently playing a game.  In particular it does *not* tell us whether Swarm was started with the `--cheat` flag enabled.

File: test/integration/Main.hs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  children <- mapM (\path -> (,) path <$> readFile path) filePaths
+  -- recurse
+  sub <- filterM doesDirectoryExist paths
+  transChildren <- concat <$> mapM (`acquire` ext) sub
+  return $ children <> transChildren
+ where
+  hasExt path = takeExtension path == ("." ++ ext)
+
+data Time
+  = -- | One second should be anough to run most programs.
```
Comment (byorgey) [inline test/integration/Main.hs:87]:
```suggestion
  = -- | One second should be enough to run most programs.
```

File: src/Swarm/Game/State.hs (hunk 1 above)
Comment (xsebek) [reply src/Swarm/Game/State.hs:525]: [Fix creative mode ignoring scenario inventory](https://github.com/swarm-game/swarm/pull/400/commits/602cefa70739883a076109495df622b17cbb9383) :slightly_smiling_face:

If I add inventory to the base in a scenario, originally it would overwrite it with zero of everything.
I could workaround by adding `program: "create X; create Y; ..."`, but I thought it was a bug, so I fixed it.

File: src/Swarm/TUI/Controller.hs (hunk 2 above)
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs:148]: Well spotted! :+1: Weird that I did not notice, but I guess I did not test it too much. :sweat_smile:

File: test/integration/Main.hs (hunk 3 above)
Comment (xsebek) [reply test/integration/Main.hs]: Thanks, fixed :+1:

File: src/Swarm/TUI/Controller.hs (hunk 2 above)
Comment (xsebek) [reply src/Swarm/TUI/Controller.hs]: Now it is `(s ^. uiState . uiCheatMode)` as it should have been. :heavy_check_mark:

File: src/Swarm/Game/State.hs (hunk 1 above)
Comment (xsebek) [reply src/Swarm/Game/State.hs:525]: The reason why I looked for it was that the new test case would not work without inventory in creative mode. 

(I was also trying modified versions of it where you start with a logger and give it to the robot.)

I expect being able to specify inventory even in creative mode will come in handy. :slightly_smiling_face:

**Comment by xsebek** [comment]
I rebased, cleaned up the history and added the new portal challenge solution to the tests :innocent: 

@byorgey given that there are some changes outside tests, I would like to merge the commits as they are now.
Unless you have more suggestions of course :slightly_smiling_face:

File: src/Swarm/Game/State.hs (hunk 1 above)
Comment (byorgey) [reply src/Swarm/Game/State.hs:525]: Aha, I see now! 👍

**Review by byorgey — APPROVED** [review]
Looks good!
